src/bootstrap/src/core/builder/mod.rs

```diff
@@ -259,12 +259,17 @@ impl PathSet {
     // internal use only
     fn check(p: &TaskPath, needle: &Path, module: Kind) -> bool {
         if let Some(p_kind) = &p.kind {
-            p.path.ends_with(needle) && *p_kind == module
+            Self::check_path(p, needle) && *p_kind == module
         } else {
-            p.path.ends_with(needle)
+            Self::check_path(p, needle)
         }
     }
 
+    fn check_path(p: &TaskPath, needle: &Path) -> bool {
+        // This order is important for retro-compatibility, as `starts_with` was introduced later.
+        p.path.ends_with(needle) || p.path.starts_with(needle)
+    }
+
```

Comment:
We will never need `check_path` outside of `check` context, so how about making `check_path` a closure inside `check` function ?